lib/internal/modules/esm/translators.js

```diff
@@ -501,19 +509,59 @@ translators.set('wasm', async function(url, source) {
     throw err;
   }
 
-  const imports =
-      ArrayPrototypeMap(WebAssembly.Module.imports(compiled),
-                        ({ module }) => module);
-  const exports =
-    ArrayPrototypeMap(WebAssembly.Module.exports(compiled),
-                      ({ name }) => name);
+  const importsList = new SafeSet();
+  const wasmGlobalImports = [];
+  for (const impt of WebAssembly.Module.imports(compiled)) {
+    if (impt.kind === 'global') {
+      ArrayPrototypePush(wasmGlobalImports, impt);
+    }
+    importsList.add(impt.module);
+  }
+
+  const exportsList = new SafeSet();
+  const wasmGlobalExports = new SafeSet();
+  for (const expt of WebAssembly.Module.exports(compiled)) {
+    if (expt.kind === 'global') {
+      wasmGlobalExports.add(expt.name);
+    }
+    exportsList.add(expt.name);
+  }
 
-  const createDynamicModule = require(
-    'internal/modules/esm/create_dynamic_module');
-  const { module } = createDynamicModule(imports, exports, url, (reflect) => {
+  const createDynamicModule = require('internal/modules/esm/create_dynamic_module');
+
+  const { module } = createDynamicModule([...importsList], [...exportsList], url, (reflect) => {
+    for (const impt of importsList) {
+      const importNs = reflect.imports[impt];
+      const wasmInstance = wasmInstances.get(importNs);
+      if (wasmInstance) {
+        const wrappedModule = ObjectAssign({ __proto__: null }, reflect.imports[impt]);
+        for (const { module, name } of wasmGlobalImports) {
+          if (module !== impt) {
+            continue;
+          }
+          // Import of Wasm module global -> get direct WebAssembly.Global wrapped value.
+          // JS API validations otherwise remain the same.
+          wrappedModule[name] = wasmInstance[name];
+        }
+        reflect.imports[impt] = wrappedModule;
+      }
+    }
+    // In cycles importing unexecuted Wasm, wasmInstance will be undefined, which will fail during
+    // instantiation, since all bindings will be in the Temporal Deadzone (TDZ).
     const { exports } = new WebAssembly.Instance(compiled, reflect.imports);
-    for (const expt of ObjectKeys(exports)) {
-      reflect.exports[expt].set(exports[expt]);
+    wasmInstances.set(module.getNamespace(), exports);
+    for (const expt of exportsList) {
+      let val = exports[expt];
+      // Unwrap WebAssembly.Global for JS bindings
+      if (wasmGlobalExports.has(expt)) {
+        try {
+          val = val.value;
+        } catch {
+          // v128 doesn't support ToJsValue() -> undefined (ideally should stay in TDZ)
+          continue;
```

Comment:
Can we avoid this try catch by checking for undefined? It would be a lot faster

Reply:
The `WebAssembly.Global` `ToJSValue` implementation will throw for `v128`, and here we are triggering that getter (https://webassembly.github.io/esm-integration/js-api/index.html#tojsvalue).

We could use the type information to know when it's a v128 to remove the try-catch, and I think that's the optimization that should happen here. But in order to do that we need to parse the Wasm binary separately to V8. I did have a small implementation of this in an earlier version of this PR, but inlining a mini wasm parser seemed like overkill even though it was possible in a few hundred lines. We should definitely follow-up on this optimization though, although ideally it could just be a first-class `WebAssemblyModuleRecord` when V8 implements instances itself.

Reply:
An alternative could just be to let the error happen and state that v128 globals are not exportable. But that seems like an unnecessary limitation to me.

Reply:
Agreed. Thanks for the explanation.